### PR TITLE
Fix segmentation fault during agent registration

### PR DESCRIPTION
**Pull request by achawla2012**
### Proposed changes

- Add nil check around feature object.
- Add unit tests during agent registration

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [X] I have read the [`CONTRIBUTING`](https://github.com/nginx/agent/blob/main/docs/CONTRIBUTING.md) document
- [X] I have run ```make install-tools``` and have attached any dependency changes to this pull request
- [X] If applicable, I have added tests that prove my fix is effective or that my feature works
- [X] If applicable, I have checked that any relevant tests pass after adding my changes
- [X] If applicable, I have updated any relevant documentation ([`README.md`](https://github.com/nginx/agent/blob/main/README.md))
- [ ] If applicable, I have tested my cross-platform changes on Ubuntu 22, Redhat 8, SUSE 15 and FreeBSD 13

### CODE REVIEW

File: src/plugins/commander.go [hunk 1]
```diff
@@ -163,11 +163,14 @@ func (c *Commander) agentRegistered(cmd *proto.Command) {
 				}
 			}
 
-			for index, feature := range agtCfg.Details.Features {
-				agtCfg.Details.Features[index] = strings.Replace(feature, "features_", "", 1)
+			if agtCfg.Details != nil && agtCfg.Details.Features != nil {
```
Comment (ryepup) [inline src/plugins/commander.go:166]: nit: it's safe to `range` over a nil slice ([demo](https://go.dev/play/p/dADYKu93P57)), we can omit one of these checks

```suggestion
			if agtCfg.Details != nil {
```

File: src/plugins/commander.go [hunk 2]
```diff
@@ -163,11 +163,14 @@ func (c *Commander) agentRegistered(cmd *proto.Command) {
...
+				for index, feature := range agtCfg.Details.Features {
+					agtCfg.Details.Features[index] = strings.Replace(feature, "features_", "", 1)
+				}
+
+				sort.Strings(agtCfg.Details.Features)
 			}
 
-			sort.Strings(agtCfg.Details.Features)
 			sort.Strings(c.config.Features)
```
Comment (achawla2012) [inline src/plugins/commander.go:174]: Need feedback on `sort.Strings(c.config.Features)`.

I have not added nil check around it with assumption `c.config.Features` won't be null.
Is my assumption true?

Comment (ryepup) [reply src/plugins/commander.go:174]: I think we're OK on this line,`sort.Strings` works fine if given a `nil`: https://go.dev/play/p/91Jt0vRD9vv

The next few lines look like more potential nil pointer derefs if `agtCfg.Details` is nil. Maybe we make one larger `if agtCfg.Details != nil {` block and pull everything down to L189 into it? Seems like this code is all about synchronizing `agtCfg.Details.Features` and `c.config.Features`, and we don't need to do any of that work if we don't have any `agtCfg.Details`.

It's a little tempting to introduce a helper methods here, so we could have something simple in `agentRegistered`:

```go
if agtCfg.Details != nil && agtCfg.Details.Features != nil {
  c.synchronizeFeatures(agtCfg.Details.Features)
}
```

Comment (achawla2012) [reply src/plugins/commander.go:174]: At first I thought the same, but agent uses `agentConfig` as a way to see if management and data plane are sync. It's likely if `agtCfg.Details.Features` is nil but `c.config.Features` not. Agent would flag such setting as un-synchronized. There fore take steps to synchronize.

If we go by above approach, synchronization would then only happen when `agtCfg.Details.Features` is non-nil.

cc @oliveromahony

Comment (ryepup) [reply src/plugins/commander.go:174]: hrm, just noticed that L141 kinda looks like it's doing a lot of the same work?

Comment (achawla2012) [reply src/plugins/commander.go:174]: Oh yeah, it's sth that is taken care of before this. I will refactor this a bit. Thanks for pointing this.

**Review by oliveromahony — APPROVED** [review]

**Review by Dean-Coakley — APPROVED** [review]

**Review by oliveromahony — APPROVED** [review]

**Review by dhurley — APPROVED** [review]

**Pull request merged**